src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -156,11 +191,94 @@ public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSynta
-                    _syntaxNodes.Add(node.Name);
+                    var result = new SyntaxResult(node.Name, SyntaxKind.MarkupTagHelperAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(result);
                 }
 
                 base.VisitMarkupTagHelperAttribute(node);
             }
+
+            public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    var transition = new SyntaxResult(node.Transition, SyntaxKind.Transition, _razorCodeDocument);
+                    _syntaxNodes.Add(transition);
+
+                    var directiveAttribute = new SyntaxResult(node.Name, SyntaxKind.MarkupTagHelperDirectiveAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(directiveAttribute);
+
+                    if (node.Colon != null)
+                    {
+                        var colon = new SyntaxResult(node.Colon, SyntaxKind.Colon, _razorCodeDocument);
+                        _syntaxNodes.Add(colon);
+                    }
+
+                    if (node.ParameterName != null)
+                    {
+                        var parameterName = new SyntaxResult(node.ParameterName, SyntaxKind.MarkupTagHelperDirectiveAttribute, _razorCodeDocument);
+                        _syntaxNodes.Add(parameterName);
+                    }
+                }
+
+                base.VisitMarkupTagHelperDirectiveAttribute(node);
+            }
+
+            public override void VisitMarkupMinimizedTagHelperDirectiveAttribute(MarkupMinimizedTagHelperDirectiveAttributeSyntax node)
+            {
+                if (node.TagHelperAttributeInfo.Bound)
+                {
+                    var transition = new SyntaxResult(node.Transition, SyntaxKind.Transition, _razorCodeDocument);
+                    _syntaxNodes.Add(transition);
+
+                    var directiveAttribute = new SyntaxResult(node.Name, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute, _razorCodeDocument);
+                    _syntaxNodes.Add(directiveAttribute);
+
+                    if (node.Colon != null)
+                    {
+                        var colon = new SyntaxResult(node.Colon, SyntaxKind.Colon, _razorCodeDocument);
+                        _syntaxNodes.Add(colon);
+                    }
+
+                    if (node.ParameterName != null)
+                    {
+                        var parameterName = new SyntaxResult(node.ParameterName, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute, _razorCodeDocument);
+                        _syntaxNodes.Add(parameterName);
+                    }
+                }
+
+                base.VisitMarkupMinimizedTagHelperDirectiveAttribute(node);
+            }
+
+            private bool ClassifyTagName(MarkupTagHelperElementSyntax node)
+            {
+                var name = node.StartTag.Name.Content;
```

Comment:
1. Will StartTag be null here? Ex: `<div></span></div>`
2. Will Name ever be null here? Ex: What happens when the tag is `<></>`?

Reply:
Looking more this probably won't happen because we are only operating on tag helper here. Wouldn't hurt to be defensive here though.